Index: upower/src/org.freedesktop.UPower.service.in
===================================================================
--- upower.orig/src/org.freedesktop.UPower.service.in	2011-10-03 17:08:20.000000000 +0200
+++ upower/src/org.freedesktop.UPower.service.in	2012-02-11 16:59:05.356308019 +0100
@@ -2,4 +2,5 @@
 Name=org.freedesktop.UPower
 Exec=@libexecdir@/upowerd
 User=root
+SystemdService=upower.service
 
