commit b8a200eb481a42adf26d639dbdc2224a6c99f841
Author: Richard Hughes <richard@hughsie.com>
Date:   Mon Jan 25 14:25:37 2010 +0000

    Make DkpClient a singleton to avoid some weird race conditions

diff --git a/devkit-power-gobject/dkp-client.c b/devkit-power-gobject/dkp-client.c
index feac826..5cef4f9 100644
--- a/devkit-power-gobject/dkp-client.c
+++ b/devkit-power-gobject/dkp-client.c
@@ -74,6 +74,7 @@ enum {
 };
 
 static guint signals [DKP_CLIENT_LAST_SIGNAL] = { 0 };
+static gpointer dkp_client_object = NULL;
 
 G_DEFINE_TYPE (DkpClient, dkp_client, G_TYPE_OBJECT)
 
@@ -734,8 +735,12 @@ dkp_client_finalize (GObject *object)
 DkpClient *
 dkp_client_new (void)
 {
-	DkpClient *client;
-	client = g_object_new (DKP_TYPE_CLIENT, NULL);
-	return DKP_CLIENT (client);
+	if (dkp_client_object != NULL) {
+		g_object_ref (dkp_client_object);
+	} else {
+		dkp_client_object = g_object_new (DKP_TYPE_CLIENT, NULL);
+		g_object_add_weak_pointer (dkp_client_object, &dkp_client_object);
+	}
+	return DKP_CLIENT (dkp_client_object);
 }
 
