Index: upower/src/Makefile.in
===================================================================
--- upower.orig/src/Makefile.in	2010-03-26 07:56:49.688691489 +0100
+++ upower/src/Makefile.in	2010-03-26 07:56:41.784688652 +0100
@@ -40,13 +40,11 @@
 @BACKEND_TYPE_DUMMY_TRUE@	dummy/libupshared.la
 
 @BACKEND_TYPE_FREEBSD_TRUE@am__append_2 = \
-@BACKEND_TYPE_FREEBSD_TRUE@	freebsd/libupshared.la					\
-@BACKEND_TYPE_FREEBSD_TRUE@	$(DEVKIT_POWER_LIBS)
+@BACKEND_TYPE_FREEBSD_TRUE@	freebsd/libupshared.la
 
 @BACKEND_TYPE_LINUX_TRUE@am__append_3 = \
 @BACKEND_TYPE_LINUX_TRUE@	linux/libupshared.la					\
-@BACKEND_TYPE_LINUX_TRUE@	$(GUDEV_LIBS)						\
-@BACKEND_TYPE_LINUX_TRUE@	$(DEVKIT_POWER_LIBS)
+@BACKEND_TYPE_LINUX_TRUE@	$(GUDEV_LIBS)
 
 @EGG_BUILD_TESTS_TRUE@check_PROGRAMS = up-self-test$(EXEEXT)
 @EGG_BUILD_TESTS_TRUE@TESTS = up-self-test$(EXEEXT)
@@ -85,8 +83,7 @@
 @EGG_BUILD_TESTS_TRUE@	$(am__objects_1)
 up_self_test_OBJECTS = $(am_up_self_test_OBJECTS)
 am__DEPENDENCIES_1 =
-@EGG_BUILD_TESTS_TRUE@up_self_test_DEPENDENCIES =  \
-@EGG_BUILD_TESTS_TRUE@	dummy/libupshared.la \
+@EGG_BUILD_TESTS_TRUE@up_self_test_DEPENDENCIES = dummy/libuptest.la \
 @EGG_BUILD_TESTS_TRUE@	$(am__DEPENDENCIES_1) \
 @EGG_BUILD_TESTS_TRUE@	$(am__DEPENDENCIES_1) \
 @EGG_BUILD_TESTS_TRUE@	$(am__DEPENDENCIES_1) \
@@ -107,11 +104,10 @@
 	$(am__objects_2)
 upowerd_OBJECTS = $(am_upowerd_OBJECTS)
 @BACKEND_TYPE_LINUX_TRUE@am__DEPENDENCIES_2 = linux/libupshared.la \
-@BACKEND_TYPE_LINUX_TRUE@	$(am__DEPENDENCIES_1) \
-@BACKEND_TYPE_LINUX_TRUE@	$(DEVKIT_POWER_LIBS)
+@BACKEND_TYPE_LINUX_TRUE@	$(am__DEPENDENCIES_1)
 upowerd_DEPENDENCIES = $(am__DEPENDENCIES_1) $(am__DEPENDENCIES_1) \
-	$(am__DEPENDENCIES_1) $(UPOWER_LIBS) $(am__append_1) \
-	$(am__append_2) $(am__DEPENDENCIES_2)
+	$(am__DEPENDENCIES_1) $(UPOWER_LIBS) $(DEVKIT_POWER_LIBS) \
+	$(am__append_1) $(am__append_2) $(am__DEPENDENCIES_2)
 upowerd_LINK = $(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) \
 	$(LIBTOOLFLAGS) --mode=link $(CCLD) $(upowerd_CFLAGS) \
 	$(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
@@ -447,8 +443,8 @@
 	$(AM_CPPFLAGS)
 
 upowerd_LDADD = -lm $(USB_LIBS) $(GIO_LIBS) $(DBUS_GLIB_LIBS) \
-	$(POLKIT_LIBS) $(UPOWER_LIBS) $(am__append_1) $(am__append_2) \
-	$(am__append_3)
+	$(POLKIT_LIBS) $(UPOWER_LIBS) $(DEVKIT_POWER_LIBS) \
+	$(am__append_1) $(am__append_2) $(am__append_3)
 upowerd_CFLAGS = \
 	$(WARNINGFLAGS_C)					\
 	$(NULL)
@@ -479,7 +475,7 @@
 
 @EGG_BUILD_TESTS_TRUE@up_self_test_LDADD = \
 @EGG_BUILD_TESTS_TRUE@	-lm							\
-@EGG_BUILD_TESTS_TRUE@	dummy/libupshared.la					\
+@EGG_BUILD_TESTS_TRUE@	dummy/libuptest.la					\
 @EGG_BUILD_TESTS_TRUE@	$(GLIB_LIBS)						\
 @EGG_BUILD_TESTS_TRUE@	$(GIO_CFLAGS)						\
 @EGG_BUILD_TESTS_TRUE@	$(DBUS_GLIB_LIBS)					\
Index: upower/src/dummy/Makefile.in
===================================================================
--- upower.orig/src/dummy/Makefile.in	2010-03-26 07:56:50.264693640 +0100
+++ upower/src/dummy/Makefile.in	2010-03-26 07:56:41.908688797 +0100
@@ -34,6 +34,8 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
+@BACKEND_TYPE_DUMMY_TRUE@am__append_1 = libupshared.la
+@EGG_BUILD_TESTS_TRUE@am__append_2 = libuptest.la
 subdir = src/dummy
 DIST_COMMON = $(srcdir)/Makefile.am $(srcdir)/Makefile.in
 ACLOCAL_M4 = $(top_srcdir)/aclocal.m4
@@ -51,6 +53,16 @@
 AM_V_lt = $(am__v_lt_$(V))
 am__v_lt_ = $(am__v_lt_$(AM_DEFAULT_VERBOSITY))
 am__v_lt_0 = --silent
+@BACKEND_TYPE_DUMMY_TRUE@am_libupshared_la_rpath =
+libuptest_la_LIBADD =
+am__libuptest_la_SOURCES_DIST = up-backend.c up-native.c
+am__objects_1 = libuptest_la-up-backend.lo libuptest_la-up-native.lo
+@EGG_BUILD_TESTS_TRUE@am_libuptest_la_OBJECTS = $(am__objects_1)
+libuptest_la_OBJECTS = $(am_libuptest_la_OBJECTS)
+libuptest_la_LINK = $(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) \
+	$(LIBTOOLFLAGS) --mode=link $(CCLD) $(libuptest_la_CFLAGS) \
+	$(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
+@EGG_BUILD_TESTS_TRUE@am_libuptest_la_rpath =
 DEFAULT_INCLUDES = -I.@am__isrc@ -I$(top_builddir)
 depcomp = $(SHELL) $(top_srcdir)/depcomp
 am__depfiles_maybe = depfiles
@@ -77,8 +89,9 @@
 AM_V_GEN = $(am__v_GEN_$(V))
 am__v_GEN_ = $(am__v_GEN_$(AM_DEFAULT_VERBOSITY))
 am__v_GEN_0 = @echo "  GEN   " $@;
-SOURCES = $(libupshared_la_SOURCES)
-DIST_SOURCES = $(libupshared_la_SOURCES)
+SOURCES = $(libupshared_la_SOURCES) $(libuptest_la_SOURCES)
+DIST_SOURCES = $(libupshared_la_SOURCES) \
+	$(am__libuptest_la_SOURCES_DIST)
 ETAGS = etags
 CTAGS = ctags
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
@@ -256,19 +269,19 @@
 INCLUDES = \
 	-I$(top_builddir)/src -I$(top_srcdir)/src		\
 	-DUP_COMPILATION					\
-	-DEGG_TEST						\
-	-I$(top_srcdir)/devkit-power-gobject			\
 	-I$(top_srcdir)/libupower-glib				\
 	$(DBUS_GLIB_CFLAGS)					\
 	$(POLKIT_CFLAGS)					\
 	$(GLIB_CFLAGS)
 
-noinst_LTLIBRARIES = libupshared.la
+noinst_LTLIBRARIES = $(am__append_1) $(am__append_2)
 libupshared_la_SOURCES = \
 	up-backend.c						\
 	up-native.c						\
 	$(BUILT_SOURCES)
 
+@EGG_BUILD_TESTS_TRUE@libuptest_la_CFLAGS = -DEGG_TEST
+@EGG_BUILD_TESTS_TRUE@libuptest_la_SOURCES = $(libupshared_la_SOURCES)
 all: all-am
 
 .SUFFIXES:
@@ -313,7 +326,9 @@
 	  rm -f "$${dir}/so_locations"; \
 	done
 libupshared.la: $(libupshared_la_OBJECTS) $(libupshared_la_DEPENDENCIES) 
-	$(AM_V_CCLD)$(LINK)  $(libupshared_la_OBJECTS) $(libupshared_la_LIBADD) $(LIBS)
+	$(AM_V_CCLD)$(LINK) $(am_libupshared_la_rpath) $(libupshared_la_OBJECTS) $(libupshared_la_LIBADD) $(LIBS)
+libuptest.la: $(libuptest_la_OBJECTS) $(libuptest_la_DEPENDENCIES) 
+	$(AM_V_CCLD)$(libuptest_la_LINK) $(am_libuptest_la_rpath) $(libuptest_la_OBJECTS) $(libuptest_la_LIBADD) $(LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
@@ -321,6 +336,8 @@
 distclean-compile:
 	-rm -f *.tab.c
 
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/libuptest_la-up-backend.Plo@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/libuptest_la-up-native.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/up-backend.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/up-native.Plo@am__quote@
 
@@ -348,6 +365,22 @@
 @AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
 @am__fastdepCC_FALSE@	$(LTCOMPILE) -c -o $@ $<
 
+libuptest_la-up-backend.lo: up-backend.c
+@am__fastdepCC_TRUE@	$(AM_V_CC)$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(libuptest_la_CFLAGS) $(CFLAGS) -MT libuptest_la-up-backend.lo -MD -MP -MF $(DEPDIR)/libuptest_la-up-backend.Tpo -c -o libuptest_la-up-backend.lo `test -f 'up-backend.c' || echo '$(srcdir)/'`up-backend.c
+@am__fastdepCC_TRUE@	$(AM_V_at)$(am__mv) $(DEPDIR)/libuptest_la-up-backend.Tpo $(DEPDIR)/libuptest_la-up-backend.Plo
+@am__fastdepCC_FALSE@	$(AM_V_CC) @AM_BACKSLASH@
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	source='up-backend.c' object='libuptest_la-up-backend.lo' libtool=yes @AMDEPBACKSLASH@
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
+@am__fastdepCC_FALSE@	$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(libuptest_la_CFLAGS) $(CFLAGS) -c -o libuptest_la-up-backend.lo `test -f 'up-backend.c' || echo '$(srcdir)/'`up-backend.c
+
+libuptest_la-up-native.lo: up-native.c
+@am__fastdepCC_TRUE@	$(AM_V_CC)$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(libuptest_la_CFLAGS) $(CFLAGS) -MT libuptest_la-up-native.lo -MD -MP -MF $(DEPDIR)/libuptest_la-up-native.Tpo -c -o libuptest_la-up-native.lo `test -f 'up-native.c' || echo '$(srcdir)/'`up-native.c
+@am__fastdepCC_TRUE@	$(AM_V_at)$(am__mv) $(DEPDIR)/libuptest_la-up-native.Tpo $(DEPDIR)/libuptest_la-up-native.Plo
+@am__fastdepCC_FALSE@	$(AM_V_CC) @AM_BACKSLASH@
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	source='up-native.c' object='libuptest_la-up-native.lo' libtool=yes @AMDEPBACKSLASH@
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
+@am__fastdepCC_FALSE@	$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(libuptest_la_CFLAGS) $(CFLAGS) -c -o libuptest_la-up-native.lo `test -f 'up-native.c' || echo '$(srcdir)/'`up-native.c
+
 mostlyclean-libtool:
 	-rm -f *.lo
 
Index: upower/src/freebsd/Makefile.in
===================================================================
--- upower.orig/src/freebsd/Makefile.in	2010-03-26 07:56:50.112694665 +0100
+++ upower/src/freebsd/Makefile.in	2010-03-26 07:56:42.064688001 +0100
@@ -264,7 +264,6 @@
 INCLUDES = \
 	-I$(top_builddir)/src -I$(top_srcdir)/src		\
 	-DUP_COMPILATION					\
-	-I$(top_srcdir)/devkit-power-gobject			\
 	-I$(top_srcdir)/libupower-glib				\
 	$(DBUS_GLIB_CFLAGS)					\
 	$(POLKIT_CFLAGS)					\
Index: upower/src/linux/Makefile.in
===================================================================
--- upower.orig/src/linux/Makefile.in	2010-03-26 07:56:49.824694569 +0100
+++ upower/src/linux/Makefile.in	2010-03-26 07:56:42.232687338 +0100
@@ -267,7 +267,6 @@
 	-I$(top_builddir)/src -I$(top_srcdir)/src		\
 	-DUP_COMPILATION					\
 	-DG_UDEV_API_IS_SUBJECT_TO_CHANGE			\
-	-I$(top_srcdir)/devkit-power-gobject			\
 	-I$(top_srcdir)/libupower-glib				\
 	$(GIO_CFLAGS)						\
 	$(DBUS_GLIB_CFLAGS)					\
